common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java

```diff
@@ -206,19 +296,79 @@ private void setSubtitleTextPre(ClientboundSetSubtitleTextPacket packet, Callbac
+            cancellable = true)
+    private void handlePlayerChat(ClientboundPlayerChatPacket packet, CallbackInfo ci) {
         if (!isRenderThread()) return;
-        ChatPacketReceivedEvent result = EventFactory.onChatReceived(chatType, message);
-        if (result.isCanceled()) return;
+        ChatPacketReceivedEvent result =
+                EventFactory.onChatReceived(com.wynntils.mc.objects.ChatType.CHAT, packet.unsignedContent());
+        if (result.isCanceled()) {
+            ci.cancel();
+            return;
+        }
+
+        if (!result.getMessage().equals(packet.unsignedContent())) {
+            // Because of the injection point, we know these optionals are present
+            SignedMessageBody signedMessageBody =
+                    packet.body().unpack(this.messageSignatureCache).get();
+            ChatType.Bound bound = packet.chatType()
+                    .resolve(this.registryAccess.compositeAccess())
+                    .get();
 
-        gui.handleChat(chatType, result.getMessage(), uuid);
+            UUID uuid = packet.sender();
+            PlayerInfo playerInfo = this.getPlayerInfo(uuid);
+            RemoteChatSession remoteChatSession = playerInfo.getChatSession();
+
+            SignedMessageLink signedMessageLink;
+            if (remoteChatSession != null) {
+                signedMessageLink = new SignedMessageLink(packet.index(), uuid, remoteChatSession.sessionId());
+            } else {
+                signedMessageLink = SignedMessageLink.unsigned(uuid);
+            }
+
+            PlayerChatMessage playerChatMessage = new PlayerChatMessage(
+                    signedMessageLink,
+                    packet.signature(),
+                    signedMessageBody,
+                    packet.unsignedContent(),
+                    packet.filterMask());
+
+            this.minecraft.getChatListener().handlePlayerChatMessage(playerChatMessage, playerInfo.getProfile(), bound);
+            this.messageSignatureCache.push(playerChatMessage);
+
+            ci.cancel();
+        }
+    }
+
+    @Inject(
+            method = "handleSystemChat",
+            at =
+                    @At(
+                            value = "INVOKE",
+                            target =
+                                    "Lnet/minecraft/client/multiplayer/chat/ChatListener;handleSystemMessage(Lnet/minecraft/network/chat/Component;Z)V"),
+            cancellable = true)
+    private void handleSystemChat(ClientboundSystemChatPacket packet, CallbackInfo ci) {
+        if (!isRenderThread()) return;
+        ChatPacketReceivedEvent result = EventFactory.onChatReceived(
+                packet.overlay() ? com.wynntils.mc.objects.ChatType.GAME_INFO : com.wynntils.mc.objects.ChatType.SYSTEM,
```

Comment:
I thought you said there were no SYSTEM? Or was it just that Wyncraft never sent any?

Reply:
There were no Chat. If you look above, the `handlePlayerChat`, that is where we would get CHAT. But we don't. I suggest we merge this as is, and do some modifications on 1.18.2, and merge it in here.

Reply:
Ok, let's make a change in 1.18 so we stop differentiate between CHAT and SYSTEM.